### PR TITLE
feat(performance): add merge-on-read performance profiling models

### PR DESCRIPTION
**Pull request by jamesbconner**
- Add QueryPerformanceProfile dataclass to capture individual query execution metrics
- Add MergeOnReadPerformance dataclass with computed properties for overhead analysis
* overhead_ms: time difference between queries with and without deletes
* overhead_percentage: relative performance impact calculation
* rows_deleted: count of rows filtered by delete files
- Add profile_query_performance method to ProfilingBackend protocol for performance measurement
- Export new performance models from models package
- Update product specification with Story 6 for performance profiling requirements
* Define acceptance criteria for merge-on-read performance measurement
* Specify timing breakdown requirements (data scan, delete scan, merge, total)
* Clarify performance profiling as optional backend capability
- Add comprehensive test coverage for performance models in test_performance_models.py

---

> [!NOTE]
> Add merge-on-read performance profiling models, extend the profiling backend with an optional performance API, update docs, and add tests.
> 
> - **Models**:
>   - Add `QueryPerformanceProfile` and `MergeOnReadPerformance` with computed properties (`overhead_ms`, `overhead_percentage`, `rows_deleted`).
>   - Export new models via `table_sleuth.models.__init__`.
> - **Profiling Backend**:
>   - Convert `ProfilingBackend` to an abstract base class; keep column profiling methods abstract.
>   - Add optional `profile_query_performance(snapshot, query, filters)` that raises `NotImplementedError` by default.
> - **Tests**:
>   - Add `tests/test_performance_models.py` covering performance model behaviors and edge cases.
>   - Update fake profiler tests to assert `NotImplementedError` for performance profiling.
> - **Docs**:
>   - Update `docs/product_specification.md` to add performance profiling to goals, define Story 6 with acceptance criteria and timing breakdown, renumber Story 7, and update success metrics.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a16eb271eb89028081bb0438f66c8ad56b95e5cb. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/table_sleuth/models/performance.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    def overhead_ms(self) -> float:
+        """Time overhead caused by merge-on-read in milliseconds."""
+        return self.with_deletes.execution_time_ms - self.without_deletes.execution_time_ms
+
+    @property
+    def overhead_percentage(self) -> float:
+        """Percentage overhead caused by merge-on-read."""
+        if self.without_deletes.execution_time_ms == 0:
+            return 0.0
+        return (self.overhead_ms / self.without_deletes.execution_time_ms) * 100
```
[the remark this answers is not in the document]
Comment (jamesbconner) [reply src/table_sleuth/models/performance.py]: bugbot review
